Fix vanilla formation selection saving

MLNW opened the pull request:
**When merged this pull request will:**

- Resolve #11 
- Add run configurations that use the launcher instead of directly launching the game

MLNW committed:
Implement logic from FixedFormations mod

One to one implementation of FixedFormations mod with some small
improvements fitting our code base.

- Bump C# version to 8.0

MLNW committed:
Add run configurations for the Bannerlord Launcher

Add the option to launch (and debug) Bannerlord via its Launcher in
order to test the mod with other mods enabled.

- Formatted the .csproj file with Rider

tbeswick96 commented on Extension/YapoSubModule.cs:
```diff
@@ -25,25 +27,48 @@ protected override void OnSubModuleLoad()
                 new Harmony("YAPO").PatchAll();
-            } catch (Exception exception)
+            }
+            catch (Exception e)
```
Why change this to `e`?
`Exception` is clearer. Personally prefer variable names that aren't shortened or abbreviated (except for LINQ predicates)

MLNW replied:
I usually agree with this but in the case of exceptions in a try-catch block I usually shorten them since lines tend to grow long with exception messages and all.

I'll revert it.

tbeswick96 commented on Extension/Fixes/Formations/FixedFormationsBehaviour.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public override void RegisterEvents()
+        {
+            // Nothing to register
```
Just empty contrsuctor `RegisterEvents() { }`

tbeswick96 commented on Extension/Fixes/Formations/FixedFormationsSaveableTypeDefiner.cs:
```diff
@@ -0,0 +1,24 @@
+    {
+        public FixedFormationsSaveableTypeDefiner() : base(13537200)
+        {
+            // Nothing to do
```
Just empty constructor `base(13537200) { }`

tbeswick96 reviewed (changes requested):
All looks fine, seems like a simple implementation so that's good.
Some formatting I'm not a fan of which I'll fix in a commit to master with a full .DotSettings file included so hopefully going forward it won't be as much of a problem

MLNW committed:
Add a setting to disable formation fix

- Rename Settings to YapoSettings
- Reword auto sort setting tooltip

MLNW committed:
Incorporate code review notes

- Fix formatting

MLNW commented:
Still have to test it on the 1.2.0 beta branch

MLNW commented:
As it stands we need to implement this differently for 1.1.1 and 1.2.0 since there was a change relating to the `Campaign` class that is used to figure out when everything has been loaded and the game actually starts.

Question now is do we support both versions or only one? Since this feature isn't really core to our mission it may be best to simply implement it for the newer version and have it go live as soon as 1.2.0 isn't beta anymore.

tbeswick96 commented:
How tricky is it to have it for both? Can it be resolved through reflection based on version to avoid having to explicitly specify the thing needed from `Campaign`?

tbeswick96 committed:
Merge branch 'master' into feature/11-fix-vanilla-formation-selection-saving

# Conflicts:
#	Extension/Global/Strings.cs
#	Extension/Settings.cs
#	Extension/YetAnotherPartyOrganiser.csproj

tbeswick96 committed:
Fix merge conflicts & revert C# 8 language features

MLNW commented:
> How tricky is it to have it for both? Can it be resolved through reflection based on version to avoid having to explicitly specify the thing needed from `Campaign`?

We would need to patch different methods on `TaleWorlds.CampaignSystem.Campaing`:
- `OnSessionInitialized --> 1.1.1`
- `OnSessionStart        --> 1.2.0`

From a quick look they do the same thing but for what ever reason they decided to rename the method.

For 1.2.0 there was also another patch added for `TaleWorlds.Core.BasicCharacterObject.DefaultFormationGroup`. Sounds like it was added for 1.2.0.

I haven't worked with reflection in C# so I'm not sure if it's possible. But it does sound feasible. `YAPO.Fixes.Formations.Patches.CampaignPatches` would need to take into account the current version and patch different methods depending on it. That should be possible with Harmonys `TargetMethod` syntax.

For the `DefaultFormationGroup` I'm not sure. It may be problematic if it was only added for 1.2.0 but there is probably a way similar to the one above.

EDIT: I found this which I will try to check for the version: https://stackoverflow.com/a/12528418

MLNW committed:
Improve fix to work with beta branch

First check to see if the fix is enabled in the settings. If so then
patch the correct methods depending on if they exist or not. Fix won't
work if they change the Campaign start method to something new again but
it should also stay stable.

tbeswick96 reviewed (approved):
Lgtm
